Fixes #22594: Allow listing upstream subs

akofink opened the pull request:
To test:

```JSON
GET /katello/api/v2/organizations/1/upstream_subscriptions?page=1&per_page=4&sort_by=quantity

# Example output
{
    "total": null,
    "subtotal": 4,
    "page": "1",
    "per_page": "4",
    "error": null,
    "search": null,
    "sort": {
        "by": "quantity",
        "order": null
    },
    "results": [
        {
            "pool_id": "c505f98c8a80e016060c4fcf8b23045d",
            "quantity": 125000,
            "start_date": "2016-03-24T04:00:00+0000",
            "end_date": "2022-01-01T04:59:59+0000",
            "contract_number": "10913844",
            "consumed": 30,
            "product_name": "Employee SKU",
            "product_id": "ES1390109",
            "subscription_id": "7883907"
        },
        ...
    ]
}
```

Verify:
- manifest refresh still functions properly

jlsherrill commented:
haven't tested any of this, but the general idea seems good to me!

jturel commented on app/lib/katello/resources/candlepin.rb:
```diff
@@ -564,40 +554,11 @@ def path(id = nil)
-            end
-          end
-        end
+      class UpstreamPool < UpstreamCandlepinResource
```
What do you think about putting things like UpstreamPool into a module? => Katello::Resources::Upstream::Whatever so we can break this large file apart a bit more

akofink replied:
Yes, and split all these classes into separate files under `resources/candlepin/` and have `resource/candlepin.rb` be base + `require`s, but I was going to do that (latter) step in a separate commit.

jturel commented on app/lib/katello/resources/candlepin/pool.rb:
```diff
@@ -0,0 +1,44 @@
+      module PoolRestClient
+        class << self
+          def find(pool_id)
+            pool_json = self.get(path(pool_id), self.default_headers).body
```
I know this is WIP, but I'm curious if leaving the HTTP client methods `get`, etc up to the person including this module is your final design?

akofink replied:
Short answer: yes. It's kind of meant to be used with the `Katello::Resources::Candlepin::CandlepinResource` classes and subclasses, which all implement the `RestClient` methods.

sean797 commented:
While this is a nice edition, I wounder if it should be added to an RH type plugin? In fact I wounder if we should move all the RH specific stuff out of Katello, especially as its useless to Debain folks and purely community users.

akofink commented:
@sean797 I agree, and we have talked about this while planning this feature. We ultimately decided that moving these RH parts out into a plugin would be nice, but adding this RH feature won't make separating all of it any less difficult.

akofink commented:
This should work end to end now. I intentionally tried to add this functionality with minimal code change in the first commit; I'm going to add some refactoring commits, but first I'm going to add tests.

akofink commented:
Currently getting this error output:

```
TypeError: no implicit conversion of String into Hash      
from /home/vagrant/.rvm/gems/ruby-2.4.1/gems/rest-client-2.0.2/lib/restclient/resource.rb:50:in `merge'
```

I found some SSL related stuff online, but I'm still debugging.

akofink commented:
Make sure to look at the changes from the two commits separately. :)

akofink commented:
Insight into the test failure (thanks rubocop):

```
Updating backend data for subscription facets

From: /home/vagrant/katello/app/lib/katello/resources/candlepin/consumer.rb @ line 139 Katello::Resources::Candlepin::Consumer.virtual_guests:

    133: def virtual_guests(uuid)
    134:   response = Candlepin::CandlepinResource.get(join_path(path(uuid), 'guests'), self.default_headers).body
    135:   ::Katello::Util::Data.array_with_indifferent_access JSON.parse(response)
    136: rescue RestClient::Exception # I changed this from `rescue`
    137:   return []
    138: rescue Exception => e # I added this for debugging the failure
 => 139:   require 'pry'; binding.pry
    140: end

[1] pry(Katello::Resources::Candlepin::Consumer)> e
=> #<VCR::Errors::UnhandledHTTPRequestError: 
...
```

So, it turns out we intended to test this code but since it rescue'd any exception, it was rescuing the VCR UnhandledHTTPRequestError, which was making the test happy. Kind of interesting and a good case for why we don't want to rescue just anything.

akofink commented on test/lib/tasks/update_subscription_facet_backend_data_test.rb:
```diff
@@ -16,6 +16,8 @@ def test_run
       Katello::Host::SubscriptionFacet.any_instance.stubs(:host).returns(@host)
       Katello::Candlepin::Consumer.any_instance.stubs(:consumer_attributes).returns(:facts => {:foo => :bar})
+      Katello::Resources::Candlepin::Consumer.stubs(:virtual_guests).returns([])
+      Katello::Resources::Candlepin::Consumer.stubs(:virtual_host).returns(nil)
```
This is the fix for the test failure. Instead of the `Katello::Resources::Candlepin::Consumer` erroring with VCR and rescuing, we're just not going to hit the API at all.

akofink replied:
I did need to record a cassette for `test/services/katello/registration_manager_test.rb`.

jturel commented on app/controllers/katello/api/v2/upstream_subscriptions_controller.rb:
```diff
@@ -0,0 +1,22 @@
+    api :GET, "/organizations/:organization_id/upstream_subscriptions", N_("List available subscriptions from the CDN")
+    param :organization_id, :number, :desc => N_("Organization ID"), :required => true
+    def index
+      render json: cp_resource.get_for_owner
```
This method blows up when the org doesn't have a manifest imported. I think we should handle it more gracefully:

```
2018-02-26T15:39:49 5d4145dc [app] [E] NoMethodError: undefined method `[]' for nil:NilClass
 | /home/vagrant/katello/app/lib/katello/resources/candlepin.rb:149:in `upstream_api_uri'
 | /home/vagrant/katello/app/lib/katello/resources/candlepin.rb:153:in `site'
 | /home/vagrant/katello/app/lib/katello/resources/candlepin.rb:117:in `resource'
 | /home/vagrant/katello/app/lib/katello/resources/candlepin.rb:165:in `upstream_owner_id'
 | /home/vagrant/katello/app/lib/katello/resources/candlepin/pool.rb:53:in `get_for_owner'
 | /home/vagrant/katello/app/controllers/katello/api/v2/upstream_subscriptions_controller.rb:13:in `index'
```

akofink replied:
Good catch! Thanks for breaking! :)

akofink replied:
@jturel updated.

```
RuntimeError: Current organization has no manifest imported.
```

jlsherrill replied:
I would also recommend against rendering pure json back to the client.  We've done this in the past and it has always bitten us, for the following reasons:

* inconsistent data types (here you have candlepin based param naming (productName instead of product_name).  
* if candlepin changes some value, we propagate that change to the client (unlikely in this case)
* harder to control what gets sent to the client (i.e. we may want to reduce what it returned to improve performance).

Instead, i would recommend we populate some model object that is a simple holder of attributes.  Then we can render this as rabl just like all the other api controllers.

jturel commented on app/controllers/katello/api/v2/upstream_subscriptions_controller.rb:
```diff
@@ -0,0 +1,22 @@
+    before_action :find_organization
+
+    resource_description do
+      description "Subscriptions management on the CDN."
```
CDN refers to cdn.redhat.com which serves actual content, but upstream here is actually the Red Hat Subscription Management platform - so I think we should be more precise

jturel commented on app/controllers/katello/api/v2/upstream_subscriptions_controller.rb:
```diff
@@ -0,0 +1,22 @@
+    end
+
+    api :GET, "/organizations/:organization_id/upstream_subscriptions", N_("List available subscriptions from the CDN")
+    param :organization_id, :number, :desc => N_("Organization ID"), :required => true
```
This works, but I think we need some more functionality. Let me know if you envisioned these things being added in a later commit btw:

The way this will be used by the UI is always (afaik) going to be in the context of the upstream consumer from which the manifest was initially exported so I think we need to accept the consumer UUID parameter here to filter by, and maybe it should even be required.  Consumer#available_subscriptions has an example. We can forward those same params to hosted RHSM.

From a performance perspective I don't think retrieving the entire dataset is going to work. We have a few options: Candlepin supports 'include' and 'exclude' params which we can use to whittle down the attributes returned in the response. We will only need a few of the attributes from the pool to render our view. There is also paging: http://www.candlepinproject.org/docs/candlepin/pagination.html page & per_page seem most relevant.

What do you think?

akofink replied:
Yeah, the design doc is pretty unclear about the output columns/pagination/etc. and which params should be able to be passed. I left it untouched because of this, but I see no reason why we can't make these decisions on this PR. I like all your suggestions! /me gets to work

akofink replied:
@jturel Could you ellaborate on the required UUID param? I don't really understand why this would be required when our API users have no concept of candlepin consumers afaik. The organization_id is the only required param here because it (potentially) contains the manifest. I am unable to find the example you mentioned in Consumer#available_subscriptions - perhaps you could share a link?

jturel replied:
Yes! My thinking is: the imported manifest comes from an upstream RHSM consumer. The new UI we are building is manipulating that very consumer - changing subscription quantities, adding, and removing them altogether. While the user is not thinking in those terms it means that all of those API calls should be within the context of that consumer. We should be able to get its UUID from the identity cert if we don't have it stored elsewhere. The user shouldn't have to provide it themselves which may have made my comment confusing...

I was referring to this code: https://github.com/Katello/katello/pull/7197/files#diff-4e53ba860fcaf2eb36f0da5a96751337R83

The reason that we need to send the UUID is because we don't actually want to show all of the pools on the account - some of them may not have any entitlements available. Passing in the UUID will give us what can actually be attached to the consumer and make it into the manifest. IOW I'm not sure we have a use case to list all of the pools on the account regardless of whether they are fully consumed or not.

Sorry that was so long - hope it helps.

akofink replied:
I think I'm doing this already: https://github.com/Katello/katello/pull/7197/commits/09e012f7a63b186a48bd0d570966dedf79ffe99f#diff-d4ada19367f2e37f2adeafa9de778bccR428

jturel replied:
Ah, I do see that. It looks like the controller is using the UpstreamPool resource rather than UpstreamConsumer and so we're getting all the pools.

re: account - yep :+1: 

@jlsherrill can you validate my previous comment before Andrew makes the change?

akofink replied:
Right, we're getting all the pools for that owner. IOW, show me all the pools I could attach to my manifest. I think that's what we want?

jturel replied:
Not quite. All the pools for the owner isn't the same as what could be attached to the consumer. Maybe some of those pools have 0 entitlements available => they can not be attached.

Giving it the consumer UUID will give us a valid list of pools that can be attached which is a subset of all pools on the account/owner.

jlsherrill replied:
If i'm reading correctly, i think that jturel is suggesting we do something like this:
https://github.com/Katello/katello/blob/master/app/lib/katello/resources/candlepin.rb#L200-L204

where we are limiting to subs available to the consumer.  I'm not 100% sure if we should be using this or not, as there was some discussion around whether we should allow you to add multiple instances of the same pool (which i think the portal allows you to do in the new ui?)  Maybe we need to compare the results of the two.

jlsherrill replied:
@jturel Yeah, and i thought i asked about the zero case specifically and was told that those should still show up (my memory could be faulty though).  Maybe we should add a task to examine the pros/cons of both of these methods and leave it as it for now?

jlsherrill replied:
partially because i think it will be more obvious once we can actually use the tool

jturel replied:
The portal would have done that before new UI too. If we want to prevent double attach we'll have to do it on our side no matter how we invoke the API as far as I can tell. Comparing what is already attached to the results from upstream...

jturel replied:
:+1: to leaving for now and adjusting as we work out the feature.

jlsherrill replied:
@akofink i do think we need to make sure to revisit though, so if you could file a card or issue, that'd be awesome!

akofink replied:
http://projects.theforeman.org/issues/22710

akofink commented:
[test katello]

akofink commented:
[test katello]

jturel commented:
@akofink nice work! ACK from me, but maybe we should get a second opinion?

jlsherrill commented on app/controllers/katello/api/v2/upstream_subscriptions_controller.rb:
```diff
@@ -0,0 +1,46 @@
+        results = presenter.from_pools(JSON.parse(response.body))
+        respond(collection: {results: results})
+      else
+        fail HttpErrors::NotFound, _('Candlepin response error: %s') % response
```
returning a NotFound here is not likely what we want

jlsherrill replied:
does this not raise an exception if the response isn't 200?

jlsherrill replied:
meaning, is this if/else even required?

akofink replied:
Oh, so just let it raise the exception. :+1:

akofink replied:
I confirmed that the API user does see the exception bubble up.

i.e. An issue with key

```
{
    "displayMessage": "Neither PUB key nor PRIV key: header too long",
    "errors": [
        "Neither PUB key nor PRIV key: header too long"
    ]
}
```

jlsherrill commented on app/presenters/katello/candlepin_pools_presenter.rb:
```diff
@@ -0,0 +1,41 @@
+                  'productName',
+                  'productId',
+                  'subscriptionId'] }
+    end
```
Maybe i'm a stickler here, but a presenter is purely for display purposes and this method doesn't make much sense in one.

akofink replied:
I was looking at it from the perspective of: these are the columns this presenter expects from candlepin and knows how to display. Where else might you suggest putting it?

jlsherrill replied:
I think moving it to a model layer that handles the backend call as well would be more ideal.  I'd prefer the controller 'call' into the model layer to fetch instances of the model rather than making those calls directly from the controller.  

For example, something like:
```
pools = Katello::UpstreamPool.fetch_pools(@organization, page_num, per_page)
respond(collection: {results: pools})
```

waldenraines commented:
@akofink I assume this will include the standard pagination and the new table configuration properties as well?

akofink commented on app/controllers/katello/api/v2/upstream_subscriptions_controller.rb:
```diff
@@ -0,0 +1,46 @@
+      N_("List available subscriptions from Red Hat Subscription Management")
+    param :organization_id, :number, :desc => N_("Organization ID"), :required => true
+    param_group :cp_pagination
+    def index
```
@waldenraines here is the API description. It supports 3 params: `page`, `per_page`, and `organization_id`. You can see that pagination is slightly different than other API endpoints, as it passes the pagination params directly to the candlepin API.

jlsherrill replied:
we should try to make it as consistent with our api as possible, even if that means translating it in the controller.

jturel commented on lib/katello/permission_creator.rb:
```diff
@@ -351,6 +351,11 @@ def subscription_permissions
+                         {
+                           'katello/api/v2/upstream_subscriptions' => [:index]
+                         },
+                         :resource_type => 'Katello::Subscription'
```
If you add a new model object for the Upstream Pool I think you'll have to change this

akofink replied:
Not to my knowledge. This resource type is just what the permission is grouped under. afaiu, permissions are only controller#action based, not based on model or db access.

akofink commented:
I think I addressed everyone's comments. You can now pass `:page, :per_page, :sort_by, :order` to this API. Unfortunately, `:full_results, :search` isn't supported as candlepin always returns full results if the pagination params aren't there. For the output, `total` isn't available (there is no `consumer/:id/pools/count candlepin endpoint), but the standard `page/per_page/subtotal` is output.

waldenraines commented:
@akofink is the table configuration (similar to #7219) going to be added in a future PR?  Is there already an issue for this?

akofink commented on app/models/katello/upstream_pool.rb:
```diff
@@ -0,0 +1,45 @@
+      end
+
+      def cp_request_params
+        { include: ['id',
```
@waldenraines This API is a bit weird as it contains no information from our database (just what upstream candlepin gives us). If we add user selectable and cacheable columns, it will be in a future feature request; however, Candlepin allows include/exclude, and while we're not exposing that to our API users, we are using it ourselves right here.

akofink commented:
@JacobCallahan This one adds tests! :)

akofink commented:
I think this PR is ready to merge. We can add SAT-134 Disconnected Toggle in a separate PR.

waldenraines commented:
> I think this PR is ready to merge. We can add SAT-134 Disconnected Toggle in a separate PR.

Any thoughts on how that will be added to this (or another) API?

akofink commented:
@waldenraines It will check that setting value, and if we're disconnected, we will raise an error, similar to the errors raised for users without a default organization or an organization without a manifest imported.

akofink commented:
[test katello]

jturel commented:
[test katello]

jlsherrill commented on app/models/katello/upstream_pool.rb:
```diff
@@ -0,0 +1,45 @@
+      end
+    end
+  end
+end
```
Thanks andrew, i like this much better!

jlsherrill commented:
[test katello]

jlsherrill commented on app/controllers/katello/api/v2/upstream_subscriptions_controller.rb:
```diff
@@ -0,0 +1,29 @@
+    api :GET, "/organizations/:organization_id/upstream_subscriptions",
+      N_("List available subscriptions from Red Hat Subscription Management")
+    param :organization_id, :number, :desc => N_("Organization ID"), :required => true
+    param_group :search, Api::V2::ApiController
```
Search isn't supported here, so we probably shouldn't include this

jlsherrill replied:
looks like api controller doesn't split out paging params and search, so maybe you can split them out so that you can just reuse the paging stuff here (while still maintaining the search group the way it is)

akofink replied:
There are other differences. Here's the param group I'm thinking, which mimics the [Candlepin docs](http://www.candlepinproject.org/docs/gutterball/reportapi.html#pagination):


```ruby
    def_param_group :cp_search do
      param :page, :number, :desc => N_("Page number, starting at 1")
      param :per_page, :number, :desc => N_("Number of results per page to return.")
      param :order, String, :desc => N_("The order to sort the results in. ['asc', 'desc'] Defaults to 'desc'.")
      param :sort_by, String, :desc => N_("The field to sort the data by. Defaults to the created date.")
    end

```

jlsherrill replied:
good from me after this fix.

akofink commented:
[test katello]

akofink commented:
Any thoughts on the react-ui failures?

akofink commented:
[test katello]

akofink commented:
[test katello]

akofink commented:
Come on Jenkins... @jturel mind giving it a try?

jturel commented:
[test katello]

waldenraines commented:
> Any thoughts on the react-ui failures?

What are you seeing?

akofink commented:
http://ci.theforeman.org/blue/organizations/jenkins/katello-pr-test/detail/katello-pr-test/787/pipeline

waldenraines commented:
> Any thoughts on the react-ui failures?

https://github.com/Katello/katello/pull/7228

waldenraines commented:
[test katello]

akofink commented:
[test katello]

akofink commented:
I've squashed the commits for Jenkins' sake

akofink commented:
Thanks everyone for the great discussion! If no one has anything else, I'm going to click the button!!
